packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -0,0 +1,338 @@
+  // to have time to call the function EpochInitializer.migrateEpochAndValidators()
+  function initializeSystem(
+    uint256 firstEpochNumber,
+    uint256 firstEpochBlock,
+    address[] memory firstElected
+  ) external onlyEpochManagerEnabler {
+    require(
+      address(registry.getAddressForOrDie(CELO_UNRELEASED_TREASURE_REGISTRY_ID)).balance > 0,
+      "CeloUnreleasedTreasury not yet funded."
+    );
+    require(
+      getCeloToken().balanceOf(registry.getAddressForOrDie(CELO_UNRELEASED_TREASURE_REGISTRY_ID)) >
+        0,
+      "CeloUnreleasedTreasury not yet funded."
+    );
+    require(!systemAlreadyInitialized(), "Epoch system already initialized");
+    require(firstEpochNumber > 0, "First epoch number must be greater than 0");
+    require(firstEpochBlock > 0, "First epoch block must be greater than 0");
+    require(
+      firstEpochBlock <= block.number,
+      "First epoch block must be less or equal than current block"
+    );
+    require(firstElected.length > 0, "First elected validators must be greater than 0");
+    firstKnownEpoch = firstEpochNumber;
+    currentEpochNumber = firstEpochNumber;
+
+    Epoch storage _currentEpoch = epochs[currentEpochNumber];
+    _currentEpoch.firstBlock = firstEpochBlock;
+    _currentEpoch.startTimestamp = block.timestamp;
+
+    elected = firstElected;
+  }
+
+  // TODO maybe "freezeEpochRewards" "prepareForNextEpoch"
+
+  /// start next epoch process.
+  /// it freezes the epochrewards at the time of execution,
+  /// and starts the distribution of the rewards.
+  function startNextEpochProcess() external nonReentrant {
+    require(systemAlreadyInitialized(), "Epoch system not initialized");
+    require(isTimeForNextEpoch(), "Epoch is not ready to start");
+    require(!isOnEpochProcess(), "Epoch process is already started");
+    epochProcessing.status = EpochProcessStatus.Started;
+
+    epochs[currentEpochNumber].rewardsBlock = block.number;
+
+    // calculate rewards
+    getEpochRewards().updateTargetVotingYield();
+
+    (
+      uint256 perValidatorReward,
+      uint256 totalRewardsVoter,
+      uint256 totalRewardsCommunity,
+      uint256 totalRewardsCarbonFund
+    ) = getEpochRewards().calculateTargetEpochRewards();
+
+    epochProcessing.perValidatorReward = perValidatorReward;
+    epochProcessing.totalRewardsVoter = totalRewardsVoter;
+    epochProcessing.totalRewardsCommunity = totalRewardsCommunity;
+    epochProcessing.totalRewardsCarbonFund = totalRewardsCarbonFund;
+
+    allocateValidatorsRewards();
+
+    emit EpochProcessingStarted(currentEpochNumber);
+  }
+
+  function finishNextEpochProcess(
+    address[] calldata groups,
+    address[] calldata lessers,
+    address[] calldata greaters
+  ) external nonReentrant {
+    require(isOnEpochProcess(), "Epoch process is not started");
+    // finalize epoch
+    // last block should be the block before and timestamp from previous block
+    epochs[currentEpochNumber].lastBlock = block.number - 1;
```

Comment:
haven't we said that we might just remove `lastBlock` as it can be computed from `firstBlock` of next epoch?

Reply:
I think the consensus was to remove the `lastBlockTimestamp` since it's not possible to get historical timestamp.

We could see about removing it if you think it's an issue, but I think that getting the result without having to make extra computation is neat.